src/ILCompiler.Compiler/src/Compiler/TypeExtensions.cs

```diff
@@ -38,5 +38,36 @@ static public MetadataType GetClosestMetadataType(this TypeDesc type)
             Debug.Assert(type is MetadataType);
             return (MetadataType)type;
         }
+
+        /// <summary>
+        /// Gets a value indicating whether the method requires a hidden instantiation argument in addition
+        /// to the formal arguments defined in the method signature.
+        /// </summary>
+        public static bool RequiresInstArg(this MethodDesc method)
+        {
+            bool result = method.IsCanonicalMethod(CanonicalFormKind.Any) &&
+                (method.HasInstantiation || method.Signature.IsStatic || method.OwningType.IsValueType);
+
+            Debug.Assert(result == (method.RequiresInstMethodDescArg() || method.RequiresInstMethodTableArg()));
+
+            return result;
+        }
+
+        public static bool RequiresInstMethodDescArg(this MethodDesc method)
+        {
+            return method.IsCanonicalMethod(CanonicalFormKind.Any) && method.HasInstantiation;
+        }
+
+        public static bool RequiresInstMethodTableArg(this MethodDesc method)
+        {
+            return method.IsCanonicalMethod(CanonicalFormKind.Any) &&
+                !method.HasInstantiation &&
+                (method.Signature.IsStatic || method.OwningType.IsValueType);
+        }
+
+        public static bool IsSharedInstantiationType(this TypeDesc type)
+        {
+            return type.IsCanonicalSubtype(CanonicalFormKind.Any) || type.IsRuntimeDeterminedSubtype;
```

Comment:
I chatted with David and while it would be nice if JIT operated on the RuntimeDetermined ones, I'll probably make it operate on normal canonical and remove this helper. RuntimeDetermined will only start coming into play when code refers to dependency nodes.